File: src/variantfoldrna/workflow/scripts/snpfold_wrapper.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
)

func transcribeRNA(seq string) string {
	return strings.ReplaceAll(seq, "T", "U")
}

func changeNucleotide(nuc byte) byte {
	nucDict := map[byte]byte{'A': 'U', 'T': 'A', 'G': 'C', 'C': 'G'}
	return nucDict[nuc]
}

func runSNPFold(seq, path, temp, mutation string) string {
	print(fmt.Sprintf("%s/snpfold", path), "-T", temp, "-seq", seq, "-mut", mutation)
	cmd := exec.Command(fmt.Sprintf("%s/snpfold", path), "-T", temp, "-seq", seq, "-mut", mutation)
	output, err := cmd.Output()
	if err != nil {
		fmt.Println("Error:", err)
		return ""
	}

	result := string(output)

	return result
}

func main() {
	// Parse command line arguments
	inFile := flag.String("i", "", "Input File")
	output := flag.String("o", "", "Output")
	flank := flag.Int("flank", 0, "Flanking length")
	temp := flag.String("temp", "", "Temp")
	path := flag.String("path", "", "Path")
	flag.Parse()

	// Open input and output files
	inFilePtr, err := os.Open(*inFile)
	if err != nil {
		fmt.Println("Error opening input file:", err)
		return
	}
	defer inFilePtr.Close()

	outFilePtr, err := os.Create(*output)
	if err != nil {
		fmt.Println("Error creating output file:", err)
		return
	}
	defer outFilePtr.Close()

	// Open error file
	errorFilePtr, err := os.Create(strings.TrimSuffix(*output, ".txt") + "_error.txt")
	if err != nil {
		fmt.Println("Error creating error file:", err)
		return
	}
	defer errorFilePtr.Close()

	scanner := bufio.NewScanner(inFilePtr)
	for scanner.Scan() {
		line := scanner.Text()

		// Skip header lines
		if !strings.HasPrefix(line, "#") {
			fields := strings.Fields(line)

			// Check for indels
			if len(fields[3]) != 1 || len(fields[4]) != 1 {
				continue
			}

			// Transcribe RNA and adjust nucleotides
			seq := fields[5] + fields[3] + fields[6]
			seq = transcribeRNA(seq)
			mutation := string(fields[3]) + strconv.Itoa(int(*flank)+1) + string(fields[4])

			// Run SNPfold
			result := runSNPFold(seq, *path, *temp, mutation)

			results := strings.Split(result, ",")

			if len(results) != 5 {
				errorFilePtr.WriteString(line + "\tNA\tNA\tNA\tNA\tNA")
				continue
			}
			// Write result to output file
			_, err := fmt.Fprintf(outFilePtr, "%s\t%s\t%s\t%s\t%s\t%s", line, results[0], results[1], results[2], results[3], results[4])
			if err != nil {
				fmt.Println("Error writing to output file:", err)
				return
			}
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading input file:", err)
	}

	fmt.Println("Process complete.")
}

File: src/variantfoldrna/workflow/scripts/snpfold_wrapper.py
################################################################################
# Wrapper for SNPfold prediction tool
#
# Author: Kobie Kirven
#
# Assmann and Bevilacqua Labs
# The Pennsylvania State University
################################################################################

# Imports
import argparse
import os
import subprocess


# -- Functions -- #
def transcribe_rna(seq):
    """
    Convert the DNA sequence to RNA
    """
    return seq.replace("T", "U")


def change_nuc(nuc):
    """
    Complement nucleotides
    """
    nuc_dict = {"A": "U", "T": "A", "G": "C", "C": "G"}
    return nuc_dict[nuc]


def run_snpfold(seq, path, temp, mutation):
    """
    Run SNPfold on the sequence.
    """
    results = subprocess.run(
        [f"{path}/../bin/snpfold", "-T", temp, "-seq", seq, "-mut", mutation],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )   
    try:
        return results.stdout.decode("utf-8")
    except:
        return "Error"


if __name__ == "__main__":

    # Get the path to the sparcs directory
    path = os.path.dirname(os.path.realpath(__file__))
    # path = "../bin"
    # Parse the arguments
    parser = argparse.ArgumentParser(description="Determine RiboSNitches")
    parser.add_argument("--i", dest="in_file", help="Input File")
    parser.add_argument("--o", dest="output", help="Output")
    parser.add_argument("--flank", dest="flank", help="Flanking length")
    parser.add_argument("--temp", dest="temp", help="Temp")
    args = parser.parse_args()

    # Open the output files
    outfile = open(args.output, "w")
    error = open(args.output[:-4] + "_error.txt", "w")

    ids, seqs = [], []
    with open(args.in_file) as fn:
        for line in fn:

            # Skip the header
            if not line.startswith("#"):
                line = line.split("\t")

                # Check to make sure we don't have any indels
                if not len(line[3]) == 1 or not len(line[4]) == 1:
                    continue

                # Change the reference and alternative alleles
                if line[3] == "T":
                    line[3] = "U"
                if line[4] == "T":
                    line[4] = "U"

                # Get the sequence and the mutation
                seq = str(line[5]) + str(line[3]) + str(line[6])
                mutation = f"{line[3]}{int(args.flank) + 1}{line[4]}"

                # Run snpfold
                results = run_snpfold(transcribe_rna(seq), path, args.temp, mutation)

                # Write the results to the output file
                try:
                    previous = "\t".join(line).strip("\n")
                    results = '\t'.join(results.split(","))
                    outfile.write(f"{previous}\t{results}\n")
                except:
                    error.write("\t".join(line) + "\n")

    # Close the output and error files
    fn.close()
    error.close()
